File: lua/problem_34.lua
local function minPalindrome(s)
    local n = #s
    local dp = {}
    for i = 1, n do
        dp[i] = {}
        for j = 1, n do dp[i][j] = 0 end
    end
    for len = 2, n do
        for i = 1, n - len + 1 do
            local j = i + len - 1
            if s:sub(i,i) == s:sub(j,j) then
                dp[i][j] = dp[i+1] and dp[i+1][j-1] or 0
            else
                local a = dp[i+1] and dp[i+1][j] or 0
                local b = dp[i][j-1] or 0
                dp[i][j] = 1 + math.min(a, b)
            end
        end
    end
    local function build(i, j)
        if i > j then return '' end
        if i == j then return s:sub(i,i) end
        if s:sub(i,i) == s:sub(j,j) then
            return s:sub(i,i) .. build(i+1, j-1) .. s:sub(j,j)
        end
        local addLeft  = s:sub(j,j) .. build(i, j-1) .. s:sub(j,j)
        local addRight = s:sub(i,i) .. build(i+1, j) .. s:sub(i,i)
        local a = dp[i+1] and dp[i+1][j] or 0
        local b = dp[i][j-1] or 0
        if a < b then return addRight end
        if b < a then return addLeft end
        return addLeft < addRight and addLeft or addRight
    end
    return build(1, n)
end

local function problem34()
    print("Problem 34:")
    assert(minPalindrome("race")   == "ecarace",  "assertion failed")
    assert(minPalindrome("google") == "elgoogle", "assertion failed")
    assert(minPalindrome("a")      == "a",        "assertion failed")
    assert(minPalindrome("ab")     == "aba",      "assertion failed")
    print("All Tests Passed")
end

return { problem34 = problem34 }

File: python/problem_34.py
def min_palindrome(s: str) -> str:
    n = len(s)
    dp = [[0] * n for _ in range(n)]
    for length in range(2, n + 1):
        for i in range(n - length + 1):
            j = i + length - 1
            if s[i] == s[j]:
                dp[i][j] = dp[i+1][j-1]
            else:
                dp[i][j] = 1 + min(dp[i+1][j], dp[i][j-1])

    def build(i: int, j: int) -> str:
        if i > j:   return ''
        if i == j:  return s[i]
        if s[i] == s[j]:
            return s[i] + build(i+1, j-1) + s[j]
        add_left  = s[j] + build(i, j-1) + s[j]
        add_right = s[i] + build(i+1, j) + s[i]
        if dp[i+1][j] < dp[i][j-1]:
            return add_right
        if dp[i][j-1] < dp[i+1][j]:
            return add_left
        return min(add_left, add_right)

    return build(0, n - 1)


def main():
    print("Problem 34:")
    assert min_palindrome("race")   == "ecarace"
    assert min_palindrome("google") == "elgoogle"
    assert min_palindrome("a")      == "a"
    assert min_palindrome("ab")     == "aba"
    print("All Tests Passed")
